include/nlohmann/json.hpp

```diff
@@ -916,6 +916,7 @@ class basic_json // NOLINT(cppcoreguidelines-special-member-functions,hicpp-spec
             default:            // LCOV_EXCL_LINE
                 JSON_ASSERT(false); // NOLINT(cert-dcl03-c,hicpp-static-assert,misc-static-assert) LCOV_EXCL_LINE
         }
+        JSON_ASSERT(m_type == val.type());
```

Comment:
In which case is it possible for this to not be true?

Reply:
See #3425.

Reply:
Sorry that I wasn't clear, I meant which case in this switch statement, if you know.

Reply:
I've observed it in the `value_t::string` case. I can't rule it out for others. `to_json` doesn't enforce the destination value type and my attempt at injecting that information into the conversion process is failing right now.

Reply:
Ah, because `other_string_t` is a different type than `string_t`.  Thanks.